rust/vectordb/src/io/object_store.rs

```diff
@@ -0,0 +1,380 @@
+use std::{
+    fmt::Formatter,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use bytes::Bytes;
+use futures::{stream::BoxStream, FutureExt, StreamExt};
+use lance::io::object_store::WrappingObjectStore;
+use object_store::{
+    path::Path, GetOptions, GetResult, ListResult, MultipartId, ObjectMeta, ObjectStore, Result,
+};
+
+use async_trait::async_trait;
+use tokio::{
+    io::{AsyncWrite, AsyncWriteExt},
+    task::JoinHandle,
+};
+
+#[derive(Debug)]
+struct MirrowingObjectStore {
+    primary: Arc<dyn ObjectStore>,
+    secondary: Arc<dyn ObjectStore>,
+}
+
+impl std::fmt::Display for MirrowingObjectStore {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, "MirrowingObjectStore")?;
+        writeln!(f, "primary:")?;
+        self.primary.fmt(f)?;
+        writeln!(f, "secondary:")?;
+        self.secondary.fmt(f)?;
+        Ok(())
+    }
+}
+
+trait IsManifest {
+    fn is_manifest(&self) -> bool;
+}
+
+impl IsManifest for Path {
+    fn is_manifest(&self) -> bool {
+        self.to_string().contains("manifest")
+    }
+}
+
+/// An object store that mirrors write to secondsry object store first
+/// and than commit to primary object store.
+///
+/// This is meant to mirrow writes to a less-durable but lower-latency
+/// store. We have primary store that is durable but slow, and a secondary
+/// store that is fast but not asdurable
+///
+/// Note: this object store does not mirror writes to *.manifest files
+#[async_trait]
+impl ObjectStore for MirrowingObjectStore {
+    async fn put(&self, location: &Path, bytes: Bytes) -> Result<()> {
+        if location.is_manifest() {
+            self.primary.put(location, bytes).await
+        } else {
+            self.secondary.put(location, bytes.clone()).await?;
+            self.primary.put(location, bytes).await?;
+            Ok(())
+        }
+    }
+
+    async fn put_multipart(
+        &self,
+        location: &Path,
+    ) -> Result<(MultipartId, Box<dyn AsyncWrite + Unpin + Send>)> {
+        if location.is_manifest() {
+            return self.primary.put_multipart(location).await;
+        }
+
+        let (id, stream) = self.secondary.put_multipart(location).await?;
+
+        let mirroring_upload = MirroringUpload::new(
+            Pin::new(stream),
+            self.primary.clone(),
+            self.secondary.clone(),
+            location.clone(),
+        );
+
+        Ok((id, Box::new(mirroring_upload)))
+    }
+
+    async fn abort_multipart(&self, location: &Path, multipart_id: &MultipartId) -> Result<()> {
+        if location.is_manifest() {
+            return self.primary.abort_multipart(location, multipart_id).await;
+        }
+
+        self.secondary.abort_multipart(location, multipart_id).await
+    }
+
+    // Reads are routed to primary only
```

Comment:
Why are reads only routed to primary? Isn't that the slow one? What's the point of mirroring if we always just read the slow one?

Reply:
the secondary isn't garueeteed to have the file as it is expected to have garbage collection and/or less durable.

We have a implementation for reading the secondary store internally here: https://github.com/lancedb/sophon/blob/main/src/phalanx/src/cache/prewarmed.rs